### PR TITLE
Path planner dev merge2

### PR DESCRIPTION
**Pull request by JordySehn**
Major Update Path Tracker Implementing Corridor MPC Pipeline

### CODE REVIEW

File: main/src/vtr_lidar/src/path_planning/mpc_path_planner2.cpp
Comment (a-krawciw) [inline main/src/vtr_lidar/src/path_planning/mpc_path_planner2.cpp:1]: Is there a reason why we need both mpc_path_planner2 and mpc_path_planner?
If they provide complementary functionality we should give them distinct names. If planner2 deprecates planner then we should just remove planner.

I see it listed as a TODO, but refactoring the parameters and defining an explicit struct return type for multi-valued return functions would be good. The code is rather hard to parse between tuning parameters and not.

File: main/src/vtr_lidar/include/vtr_lidar/modules/planning/change_detection_module_v3.hpp
Comment (a-krawciw) [inline main/src/vtr_lidar/include/vtr_lidar/modules/planning/change_detection_module_v3.hpp:1]: Similar comment on naming. Does this v3 deprecate v2 (and ideally v1)? If so, let's remove them.

File: main/src/vtr_lidar/src/modules/odometry/odometry_icp_module.cpp [hunk 1]
```diff
@@ -498,13 +498,20 @@ void OdometryICPModule::run_(QueryCache &qdata0, OutputCache &,
...
-#endif
+//#if 1
+//    CLOG(WARNING, "lidar.odometry_icp") << "T_m_r is: " << qdata.T_r_m_odo->inverse().vec().transpose();
+//    CLOG(WARNING, "lidar.odometry_icp") << "w_m_r_in_r is: " << qdata.w_m_r_in_r_odo->transpose();
+//#endif
     //
     /// \todo double check validity when no vertex has been created
     *qdata.T_r_v_odo = T_r_m_icp * sliding_map_odo.T_vertex_this().inverse();
+
+    // Temporary modification by Jordy to test calibration of hte grizzly controller
```
Comment (a-krawciw) [inline main/src/vtr_lidar/src/modules/odometry/odometry_icp_module.cpp:509]: Is the modification temporary? Should this be removed?

File: main/src/vtr_lidar/src/modules/preprocessing/preprocessing_module_v2.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ void PreprocessingModuleV2::run_(QueryCache &qdata0, OutputCache &,
   if (config_->visualize) {
     PointCloudMsg pc2_msg;
     pcl::toROSMsg(*filtered_point_cloud, pc2_msg);
-    pc2_msg.header.frame_id = "lidar";
+    pc2_msg.header.frame_id = "os_sensor"; //"lidar" for honeycomb
```
Comment (a-krawciw) [inline main/src/vtr_lidar/src/modules/preprocessing/preprocessing_module_v2.cpp:108]: This looks like a DRDC leftover. This should probably be lidar again.

File: main/src/vtr_lidar/src/path_planning/cbit.cpp [hunk 3]
```diff
@@ -296,14 +304,14 @@ auto LidarCBIT::computeCommand(RobotState& robot_state0) -> Command {
     w_p_r_in_r(3) = 0.0;
     w_p_r_in_r(4) = 0.0;
     w_p_r_in_r(5) = -1* vel_history.back()[1];
-    CLOG(DEBUG, "mpc_debug.cbit") << "Robot velocity Used for Extrapolation: " << -w_p_r_in_r.transpose() << std::endl;
+    CLOG(ERROR, "mpc_debug.cbit") << "Robot velocity Used for Extrapolation: " << -w_p_r_in_r.transpose() << std::endl;
```
Comment (a-krawciw) [inline main/src/vtr_lidar/src/path_planning/cbit.cpp:307]: These don't look like errors, I think this should be reverted to DEBUG.

File: main/src/vtr_lidar/src/path_planning/cbit.cpp [hunk 4]
```diff
@@ -326,55 +336,213 @@ auto LidarCBIT::computeCommand(RobotState& robot_state0) -> Command {
...
+    robot_poses.push_back(T_w_p * T_p_r);
 
     // Send the robot poses and mpc prediction to rviz
-    visualize(stamp, T_w_p, T_p_r, T_p_r_extp, T_p_r_extp2, mpc_poses, robot_poses);
+    visualize(stamp, T_w_p, T_p_r, T_p_r_extp, T_p_r_extp2, mpc_poses, robot_poses, ref_pose_vec1, ref_pose_vec2);
 
     // return the computed velocity command for the first time step
     Command command;
-    command.linear.x = saturated_vel(0);
+    command.linear.x = saturated_vel(0) * 1.1; // * 1.1 added to compensate for bad grizzly internal controller set points
```
Comment (a-krawciw) [inline main/src/vtr_lidar/src/path_planning/cbit.cpp:524]: This should probably be a config parameter to remind us it's there. Kind of janky hardcode...

File: main/src/vtr_navigation/CMakeLists.txt
Comment (a-krawciw) [inline main/src/vtr_navigation/CMakeLists.txt:1]: Can the matplotlib dependency not be packaged into the exported dependencies of MPC, or wherever it's used directly?

File: main/src/vtr_path_planning/CMakeLists.txt [hunk 5]
```diff
@@ -51,6 +51,45 @@ ament_target_dependencies(${PROJECT_NAME}_cbit
   lgmath steam
 )
 
+# Hacky way of declaring the helper scripts
+file(GLOB_RECURSE SRC src/cbit/generate_pq.cpp)
+
+file(GLOB_RECURSE SRC src/cbit/utils.cpp)
+
+file(GLOB_RECURSE SRC src/cbit/cbit_path_planner.cpp)
```
Comment (a-krawciw) [inline main/src/vtr_path_planning/CMakeLists.txt:59]: Does this do what you want? Isn't the file variable SRC overwritten each time?

File: main/src/vtr_path_planning/include/vtr_path_planning/cbit/cbit_path_planner.hpp [hunk 6]
```diff
@@ -77,7 +78,7 @@ class CBITPlanner {
         // Costmap pointer
         std::shared_ptr<CBITCostmap> cbit_costmap_ptr;
 
-        CBITPlanner(CBITConfig conf_in, std::shared_ptr<CBITPath> path_in, vtr::path_planning::BasePathPlanner::RobotState& robot_state, std::shared_ptr<std::vector<Pose>> path_ptr, std::shared_ptr<CBITCostmap> costmap_ptr);
+        CBITPlanner(CBITConfig conf_in, std::shared_ptr<CBITPath> path_in, vtr::path_planning::BasePathPlanner::RobotState& robot_state, std::shared_ptr<std::vector<Pose>> path_ptr, std::shared_ptr<CBITCostmap> costmap_ptr, std::shared_ptr<CBITCorridor> corridor_ptr, double path_direction);
```
Comment (a-krawciw) [inline main/src/vtr_path_planning/include/vtr_path_planning/cbit/cbit_path_planner.hpp:81]: Could the path direction be moved into an Enum to make more readable, rather than a double?

File: main/src/vtr_lidar/include/vtr_lidar/path_planning/mpc_path_planner2.hpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+// Helper function for generating reference measurements poses from a discrete path to use for tracking the path at a desired forward velocity
+struct meas_result GenerateReferenceMeas2(std::shared_ptr<std::vector<Pose>> cbit_path_ptr,  std::tuple<double, double, double, double, double, double> robot_pose, int K, double DT, double VF);
+
+// Helper function for generating reference measurements poses from a discrete path to use for tracking the path at a desired forward velocity
+struct meas_result2 GenerateReferenceMeas3(std::shared_ptr<CBITPath> global_path_ptr, std::shared_ptr<CBITCorridor> corridor_ptr, std::tuple<double, double, double, double, double, double> robot_pose, int K, double DT, double VF, int current_sid);
```
Comment (a-krawciw) [inline main/src/vtr_lidar/include/vtr_lidar/path_planning/mpc_path_planner2.hpp:65]: Why do these functions have different names?
Are they not overloading of the same function?
C++ will automatically which version you want based on the parameters. 

If not, please name them based on their differences,

File: main/src/vtr_lidar/src/modules/localization/localization_icp_module.cpp [hunk 8]
```diff
@@ -329,6 +329,7 @@ void LocalizationICPModule::run_(QueryCache &qdata0, OutputCache &,
     *qdata.T_r_v_loc = T_r_v_icp;
     // set success
     *qdata.loc_success = true;
+    CLOG(WARNING, "lidar.localization_icp") << "The resulting transform is:" << T_r_v_icp;
```
Comment (a-krawciw) [inline main/src/vtr_lidar/src/modules/localization/localization_icp_module.cpp:332]: This feels like an INFO, but might be part of a larger required inspection of log messages repo wide.

File: main/src/vtr_path_planning/include/vtr_path_planning/mpc/custom_loss_functions.hpp
Comment (a-krawciw) [inline main/src/vtr_path_planning/include/vtr_path_planning/mpc/custom_loss_functions.hpp:1]: If this is useful, it should probably be in steam proper.

File: main/src/vtr_path_planning/include/vtr_path_planning/mpc/custom_steam_evaluators.hpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+ * \file custom_steam_evaluators.hpp
+ * \author Jordy Sehn, Autonomous Space Robotics Lab (ASRL)
+ */
+
+#pragma once
+
+#include "steam/evaluable/state_var.hpp"
+
+namespace steam {
+namespace stereo {
```
Comment (a-krawciw) [inline main/src/vtr_path_planning/include/vtr_path_planning/mpc/custom_steam_evaluators.hpp:25]: Based on how you appear to be using this, I don't think it should be in the namespace stereo anymore.

File: main/src/vtr_path_planning/src/mpc/lateral_error_evaluator.cpp
Comment (a-krawciw) [inline main/src/vtr_path_planning/src/mpc/lateral_error_evaluator.cpp:1]: The hpp and cpp file names should match. I would probably refactor the classes to be called LateralErrorEvaluatorRight etc. and change the header to lateral_error_evaluator.hpp

File: robots/src/utias_grizzly_description/urdf/accessories/honeycomb.urdf.xacro
[the remark this answers is not in the document]
Comment (a-krawciw) [reply robots/src/utias_grizzly_description/urdf/accessories/honeycomb.urdf.xacro:1]: Completed.

File: main/src/vtr_radar/src/modules/odometry/odometry_icp_module.cpp
[the remark this answers is not in the document]
Comment (a-krawciw) [reply main/src/vtr_radar/src/modules/odometry/odometry_icp_module.cpp:1]: Reverted

File: main/src/vtr_lidar/src/modules/odometry/odometry_icp_module.cpp (hunk 1 above)
Comment (JordySehn) [reply main/src/vtr_lidar/src/modules/odometry/odometry_icp_module.cpp:509]: temporary modification, but we will need it for commissioning the new robot - I would leave as is for now and I can remove in my cleanup after you and Marija merge and I branch off of this.

File: main/src/vtr_lidar/include/vtr_lidar/path_planning/mpc_path_planner2.hpp (hunk 7 above)
Comment (JordySehn) [reply main/src/vtr_lidar/include/vtr_lidar/path_planning/mpc_path_planner2.hpp:65]: They are distinct versions using different algorithms I was testing - will remove all but the best one after I branch from the main post your merge today

File: main/src/vtr_lidar/src/path_planning/cbit.cpp (hunk 3 above)
Comment (JordySehn) [reply main/src/vtr_lidar/src/path_planning/cbit.cpp:307]: correct, missed these in my cleanup - resolved in latest commit

File: main/src/vtr_lidar/src/path_planning/cbit.cpp (hunk 4 above)
Comment (JordySehn) [reply main/src/vtr_lidar/src/path_planning/cbit.cpp:524]: agreed, adding a param for it in latest commit

File: main/src/vtr_path_planning/include/vtr_path_planning/mpc/custom_steam_evaluators.hpp (hunk 9 above)
Comment (JordySehn) [reply main/src/vtr_path_planning/include/vtr_path_planning/mpc/custom_steam_evaluators.hpp:25]: fair, this was just legacy of the code I used as a template to write these evaluators - removed stereo namespace and updated relevant code in latest commit

File: main/src/vtr_lidar/include/vtr_lidar/modules/planning/change_detection_module_v3.hpp
Comment (JordySehn) [reply main/src/vtr_lidar/include/vtr_lidar/modules/planning/change_detection_module_v3.hpp:1]: yuchen had left these - Im not really sure what the differences are but Ive only every used v3. I think its save to remove the older ones and ive done so in latest commit

File: main/src/vtr_path_planning/src/mpc/lateral_error_evaluator.cpp
Comment (JordySehn) [reply main/src/vtr_path_planning/src/mpc/lateral_error_evaluator.cpp:1]: resolved in latest commit

File: main/src/vtr_path_planning/include/vtr_path_planning/cbit/cbit_path_planner.hpp (hunk 6 above)
Comment (JordySehn) [reply main/src/vtr_path_planning/include/vtr_path_planning/cbit/cbit_path_planner.hpp:81]: switched to enum in latest commit - resolved

File: main/src/vtr_lidar/src/path_planning/mpc_path_planner2.cpp
Comment (a-krawciw) [reply main/src/vtr_lidar/src/path_planning/mpc_path_planner2.cpp:1]: Marking as a todo

**Review by a-krawciw — APPROVED** [review]

File: main/src/vtr_lidar/include/vtr_lidar/path_planning/mpc_path_planner2.hpp (hunk 7 above)
Comment (a-krawciw) [reply main/src/vtr_lidar/include/vtr_lidar/path_planning/mpc_path_planner2.hpp:65]: Marking as TODO

File: main/src/vtr_lidar/src/modules/odometry/odometry_icp_module.cpp (hunk 1 above)
Comment (a-krawciw) [reply main/src/vtr_lidar/src/modules/odometry/odometry_icp_module.cpp:509]: Marking as TODO
